### PR TITLE
docs: Edge Function docs

### PR DESCRIPTION
**Pull request by kiwicopple**
Updating the Edge Function docs - general kaizen and making them more accessible

### CODE REVIEW

File: apps/docs/components/Navigation/NavigationMenu/NavigationMenu.constants.ts [hunk 1]
```diff
@@ -824,60 +824,55 @@ export const functions: NavMenuConstant = {
...
         {
-          name: 'Connecting directly to Postgres',
-          url: '/guides/functions/connect-to-postgres',
+          name: 'Integrating with Supabase Storage',
+          url: '/guides/functions/storage-caching',
         },
-        { name: 'Testing your Edge Functions', url: '/guides/functions/unit-test' },
-        { name: 'Regional Invocation', url: '/guides/functions/regional-invocation' },
-        { name: 'Troubleshooting', url: '/guides/functions/troubleshooting' },
+        { name: 'Regional invocations', url: '/guides/functions/regional-invocation' },
```
Comment (laktek) [inline apps/docs/components/Navigation/NavigationMenu/NavigationMenu.constants.ts:858]: nit: Page title for it is still `Regional Invocation`. Let's change that as well?

Comment (kiwicopple) [reply apps/docs/components/Navigation/NavigationMenu/NavigationMenu.constants.ts:858]: done 👍

**Comment by saltcod** [comment]
looks like you didn't touch this file so not in the pr, but 

> It is TypeScript first and supports WASM.

should be TypeScript-first? 


screenshot-2023-11-06-at-11 57 34

**Comment by saltcod** [comment]
https://docs-git-docs-edge-functions-supabase.vercel.app/docs/guides/functions/regional-invocation#what-happens-if-the-region-has-an-outage

maybe also: 

"... you should consider temporarily changing regions during the outage"  ?

File: apps/docs/pages/guides/functions/auth.mdx [hunk 2]
```diff
@@ -4,60 +4,91 @@ export const meta = {
...
+  const supabaseClient = createClient(
+    Deno.env.get('SUPABASE_URL') ?? '',
+    Deno.env.get('SUPABASE_ANON_KEY') ?? '',
+    { global: { headers: { Authorization: authHeader } } }
+  )
+
+})
+```
+
+Importantly, this is done _inside_ the `serve()` function, so that the Authorization header is set for each request.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/auth.mdx:32]:
```suggestion
Importantly, this is done _inside_ the `Deno.serve()` callback argument, so that the Authorization header is set for each request.
```

File: apps/docs/pages/guides/functions/auth.mdx [hunk 3]
```diff
@@ -4,60 +4,91 @@ export const meta = {
...
 
-By creating a supabase client with the auth context from the function, you can do two things:
+## Fetching the user
 
-1. Get the user object.
-2. Run queries in the context of the user with [Row Level Security (RLS)](/docs/guides/auth/row-level-security) policies enforced.
+After initializing a Supabase client with the Auth context, we can use `getUser()` to fetch the user object, and run queries in the context of the user with [Row Level Security (RLS)](/docs/guides/auth/row-level-security) policies enforced.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/auth.mdx:36]:
```suggestion
After initializing a Supabase client with the Auth context, you can use `getUser()` to fetch the user object, and run queries in the context of the user with [Row Level Security (RLS)](/docs/guides/auth/row-level-security) policies enforced.
```

File: apps/docs/pages/guides/functions/auth.mdx [hunk 4]
```diff
@@ -4,60 +4,91 @@ export const meta = {
...
 
-```js mark=14,17:19,22:23 supabase/functions/select-from-table-with-auth-rls/index.ts
+```js mark=13:14
 import { serve } from 'https://deno.land/std@0.177.0/http/server.ts'
 import { createClient } from 'https://esm.sh/@supabase/supabase-js@2'
 
 serve(async (req: Request) => {
```
Comment (charislam) [inline apps/docs/pages/guides/functions/auth.mdx:42]:
```suggestion
import { createClient } from 'https://esm.sh/@supabase/supabase-js@2'

Deno.serve(async (req: Request) => {
```

Not part of this PR I know, but need to fix it at the same time if we're changing above

File: apps/docs/pages/guides/functions/auth.mdx [hunk 5]
```diff
@@ -4,60 +4,91 @@ export const meta = {
...
-    })
-  }
+
+  const supabaseClient = createClient(
+    Deno.env.get('SUPABASE_URL') ?? '',
+    Deno.env.get('SUPABASE_ANON_KEY') ?? '',
+    { global: { headers: { Authorization: req.headers.get('Authorization')! } } }
+  )
+
+  // Now we can get the session or user object
```
Comment (charislam) [inline apps/docs/pages/guides/functions/auth.mdx:50]:
```suggestion
  // Get the session or user object
```

File: apps/docs/pages/guides/functions/auth.mdx [hunk 6]
```diff
@@ -4,60 +4,91 @@ export const meta = {
...
-See the [example on GitHub](https://github.com/supabase/supabase/blob/master/examples/edge-functions/supabase/functions/select-from-table-with-auth-rls/index.ts).
+## Row Level Security
+
+After initializing a Supabase client with the Auth context, all queries will be executed with the context of the user. For database queries, this means [Row Level Security](/docs/guides/auth/row-level-security) will be enforced.
+
+```js mark=13
+import { serve } from 'https://deno.land/std@0.177.0/http/server.ts'
+import { createClient } from 'https://esm.sh/@supabase/supabase-js@2'
+
+serve(async (req: Request) => {
```
Comment (charislam) [inline apps/docs/pages/guides/functions/auth.mdx:70]:
```suggestion
import { createClient } from 'https://esm.sh/@supabase/supabase-js@2'

Deno.serve(async (req: Request) => {
```

File: apps/docs/pages/guides/functions/cicd-workflow.mdx [hunk 7]
```diff
@@ -36,7 +41,7 @@ jobs:
       - run: supabase functions deploy --project-ref $PROJECT_ID
 ```
 
-Since Supabase CLI [v1.62.0](https://github.com/supabase/cli/releases/tag/v1.62.0) you can deploy all functions with a single command.
+## Decalaritive Configuration
```
Comment (charislam) [inline apps/docs/pages/guides/functions/cicd-workflow.mdx:44]:
```suggestion
## Declarative Configuration
```

File: apps/docs/pages/guides/functions/connect-to-postgres.mdx [hunk 8]
```diff
@@ -4,19 +4,49 @@ export const meta = {
...
 
-<div class="video-container">
-  <iframe
-    src="https://www.youtube-nocookie.com/embed/cl7EuF1-RsY"
-    frameBorder="1"
-    allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
-    allowFullScreen
-  ></iframe>
-</div>
+There are various ways to connect to your Postgres database using Edge Functions.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/connect-to-postgres.mdx:11]:
```suggestion
Connect to your Postgres database from an Edge Function by using the `supabase-js` client.
You can also substitute any other Postgres client.
```

File: apps/docs/pages/guides/functions/connect-to-postgres.mdx [hunk 9]
```diff
@@ -4,19 +4,49 @@ export const meta = {
...
+## Using supabase-js
 
-Supabase Edge Functions allow you to go beyond HTTP and can connect to your Postgres Database directly!
+The `supabase-js` client is a great option for connecting to your Supabase database since it handles authorization with Row Level Security, and it automatically formats your response as JSON.
+
+```ts index.ts
+import { serve } from 'https://deno.land/std/http/server.ts'
+import { createClient } from 'https://esm.sh/@supabase/supabase-js'
+
+serve(async (_req) => {
```
Comment (charislam) [inline apps/docs/pages/guides/functions/connect-to-postgres.mdx:21]:
```suggestion
import { createClient } from 'https://esm.sh/@supabase/supabase-js'

Deno.serve(async (_req) => {
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 10]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+### Debugging locally
+
 When [developing locally](/docs/guides/functions/local-development) you will see error messages and console log statements printed to your local terminal window.
 
+## Limitations
+
+### Deno Deploy limitations
+
+- Deno does not support outgoing connections to ports `25`, `465`, and `587`.
+- Cannot read or write to File System
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:32]:
```suggestion
- Deno-deployed functions cannot read or write to the file system.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 11]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+
+### Supabase Edge Functions
+
+- Serving of HTML content is not supported (`GET` requests that return `text/html` will be rewritten to `text/plain`).
+
+## Troubleshooting
+
+### Unable to call Edge Function
+
+If you're unable to call your Edge Function or are experiencing any CORS issues, even though you followed the [CORS guide](/docs/guides/functions/cors), please head over to [supabase.com/dashboard/project/\_/functions](https://supabase.com/dashboard/project/_/functions), select your function from the list & click Logs. Do you see any errors listed there?
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:43]:
```suggestion
If you're unable to call your Edge Function or are experiencing any CORS issues:

- Make sure you followed the [CORS guide](/docs/guides/functions/cors).
- Check your function logs. Navigate to the [Functions section in your dashboard](https://supabase.com/dashboard/project/_/functions), select your function from the list, and click `Logs`. Check for any errors.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 12]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+
+There are two debugging tools available: Invocations and Logs. Invocations shows the Request and Response for each execution, while Logs shows any platform events, including deployments and errors.
+
+### Unable to deploy Edge Function
+
+- Make sure you're on the latest version of the [Supabase CLI](/docs/guides/cli#updates).
+- Run the deploy command with the `--debug` flag.
+- Run the deploy command with the `--legacy-bundle` flag and see if that works.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:51]:
```suggestion
- Run the deploy command with the `--legacy-bundle` flag.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 13]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+- If the output from the commands above does not help you to resolve the issue, please open a support ticket via the Supabase Dashboard (by clicking the "Help" button at the top right) and include all output from the commands mentioned above.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:52]:
```suggestion
- If the output from the commands above does not help you to resolve the issue, open a support ticket via the Supabase Dashboard (by clicking the "Help" button at the top right) and include all output from the commands mentioned above.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 14]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+
+### Edge Function takes too long to respond
+
+- Head over to [supabase.com/dashboard/project/\_/functions](https://supabase.com/dashboard/project/_/functions), select your function from the list & click Logs.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:56]:
```suggestion
- Navigate to the [Functions section in your Supabase dashboard](https://supabase.com/dashboard/project/_/functions), select your function from the list, and click `Logs`.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 15]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+- In the logs, look for the `booted` event and check if they have consistent boot times.
+  - If the boot times are similar, it's likely an issue with your function's code, like a large dependency etc.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:58]:
```suggestion
  - If the boot times are similar, it's likely an issue with your function's code, such as a large dependency.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 16]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+  - If only some of the `booted` events are slow, please find the affected `region` in the metadata and submit a support request via the "Help" button at the top.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:59]:
```suggestion
  - If only some of the `booted` events are slow, find the affected `region` in the metadata and submit a support request via the "Help" button at the top.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 17]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+- Run the serve command with the `--debug` flag.
+- Support engineers can then try to run the provided sample code locally and see if they can reproduce the issue.
+- Search the [Edge Runtime](https://github.com/supabase/edge-runtime) and [CLI](https://github.com/supabase/cli) repos for the error message, to see if it has been reported before.
+- If the output from the commands above does not help you to resolve the issue, please open a support ticket via the Supabase Dashboard (by clicking the "Help" button at the top right) and include all output and details about your commands.
+
+## Advanced Techniques
+
+### Checking Function Boot Time
+
+Check the logs for the function. In the logs, look for a "Booted" event and note the reported boot time. If available, click on the event to access more details, including the regions from where the function was served. Investigate if the boot time is excessively high (`higher than 1 second`) and note any patterns or regions where it occurs.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:73]:
```suggestion
Check the logs for the function. In the logs, look for a "Booted" event and note the reported boot time. If available, click on the event to access more details, including the regions from where the function was served. Investigate if the boot time is excessively high (longer than 1 second) and note any patterns or regions where it occurs.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 18]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+
+### Finding Bundle Size
+
+To find the bundle size of a function, run the following command locally:
+
+```bash
+deno info /path/to/function/index.ts
+```
+
+Look for the "size" field in the output which represents an approximated the bundle size of the function.You can find the accurate bundle size when you deploy your function via Supabase CLI. If the function is part of a larger application, consider examining the bundle size of the specific function independently.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:83]:
```suggestion
Look for the "size" field in the output which represents the approximate bundle size of the function.You can find the accurate bundle size when you deploy your function via Supabase CLI. If the function is part of a larger application, consider examining the bundle size of the specific function independently.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 19]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+
+### Analyze Dependencies
+
+Review the dependencies listed in the output of the `deno info` command. Pay attention to any significantly large dependencies, as they can contribute to increased bundle size and potential boot time issues.
+Examine if there are any unnecessary or redundant dependencies that can be removed. Check for outdated dependencies and recommend updating to the latest versions if applicable. When running deno info make sure to provide the correct path of the import map if you use one.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:88]:
```suggestion
Examine if there are any unnecessary or redundant dependencies that can be removed. Check for outdated dependencies and update to the latest versions if possible.
```

File: apps/docs/pages/guides/functions/debugging.mdx [hunk 20]
```diff
@@ -3,18 +3,94 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
+
+```bash
+deno info /path/to/function/index.ts
+```
+
+Look for the "size" field in the output which represents an approximated the bundle size of the function.You can find the accurate bundle size when you deploy your function via Supabase CLI. If the function is part of a larger application, consider examining the bundle size of the specific function independently.
+
+### Analyze Dependencies
+
+Review the dependencies listed in the output of the `deno info` command. Pay attention to any significantly large dependencies, as they can contribute to increased bundle size and potential boot time issues.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/debugging.mdx:87]:
```suggestion
Run `deno info`, providing the path to your input map if you use one.
Review the dependencies listed in the output. Pay attention to any significantly large dependencies, as they can contribute to increased bundle size and potential boot time issues.
```

File: apps/docs/pages/guides/functions/deploy.mdx [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+  description: 'Deploy your Edge Functions to your remote Supabase Project.',
+  subtitle: 'Deploy your Edge Functions to your remote Supabase Project.',
+  tocVideo: 'rzglqRdZUQE',
+}
+
+Once you have developed your Edge Functions locally, you can deploy them to your Supabase project.
+
+## Login to the CLI
+
+If you're not yet logged into the Supabase CLI, you can do so by running:
```
Comment (charislam) [inline apps/docs/pages/guides/functions/deploy.mdx:15]:
```suggestion
Log in to the Supabase CLI if necessary:
```

File: apps/docs/pages/guides/functions/deploy.mdx [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+
+```bash
+supabase login
+```
+
+<Admonition type="tip" label="CLI not installed?">
+
+Check out the [CLI Docs](/docs/guides/cli) to learn how to install the Supabase CLI on your local machine.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/deploy.mdx:23]:
```suggestion
See the [CLI Docs](/docs/guides/cli) to learn how to install the Supabase CLI on your local machine.
```

File: apps/docs/pages/guides/functions/local-development.mdx [hunk 23]
```diff
@@ -2,14 +2,45 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
 
 You can follow the [Deno guide](https://deno.com/manual@v1.32.5/getting_started/setup_your_environment) for setting up your development environment with your favorite editor/IDE.
 
+## Deno with Visual Studio Code
+
+When using VSCode, you should install both the Deno CLI and the the Deno language server [via this link](vscode:extension/denoland.vscode-deno) or by browsing the extensions in vscode and choosing to install the _Deno_ extension.
+
+## Deno support in subfolders
+
+In a given workspace (or workspace folder), sub-paths can be enabled for Deno, while code outside those paths will be not be enabled and the vscode built-in JavaScript/TypeScript language server will be used.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/local-development.mdx:22]:
```suggestion
You can enable the Deno language server for specific sub-paths in a workspace, while using VSCode's built-in JavaScript/TypeScript language server for all other files.
```

File: apps/docs/pages/guides/functions/quickstart.mdx [hunk 24]
```diff
@@ -2,32 +2,27 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
-  sidebar_label: 'Quickstart',
-  video: 'https://www.youtube.com/v/rzglqRdZUQE',
+  title: 'Developing Edge Functions locally',
+  description: 'Get started with Edge Functions on your local machine.',
+  subtitle: 'Get started with Edge Functions on your local machine.',
+  tocVideo: 'rzglqRdZUQE',
 }
 
-Learn how to build an Edge Function locally and deploy it to the Supabase Platform in less than 7 minutes.
+Let's create a basic Edge Function on our local machine and then invoke it using the Supabase CLI.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/quickstart.mdx:11]:
```suggestion
Let's create a basic Edge Function on your local machine and then invoke it using the Supabase CLI.
```

File: apps/docs/pages/guides/functions/quickstart.mdx [hunk 25]
```diff
@@ -37,195 +32,99 @@ Let's create a new Edge Function called `hello-world` inside your project:
...
+Run `supabase status` to see your local credentials.
 
-This error handling mechanism ensures that your application can gracefully handle errors that might arise during the execution of Edge Functions, providing a more robust user experience.
+</Admonition>
 
-## Importing Node npm modules
+You should see the response `{ "message":"Hello Functions!" }`.
 
-We recommend using [esm.sh](https://esm.sh/) for importing Node.js modules that are published to npm. To do so you simply put `https://esm.sh/` in front of the package name.
+If you execute the function with a different payload the response will change.
```
Comment (charislam) [inline apps/docs/pages/guides/functions/quickstart.mdx:89]:
```suggestion
If you execute the function with a different payload, the response will change.
```

File: apps/docs/pages/guides/functions/quickstart.mdx [hunk 26]
```diff
@@ -37,195 +32,99 @@ Let's create a new Edge Function called `hello-world` inside your project:
...
 
-For example, when you want to use supabase-js within Supabase Edge Functions, you would import `createClient` as follows:
+Modify the `--data '{"name":"Functions"}'` line to `--data '{"name":"World"}'` and try invoking the command again!
```
Comment (charislam) [inline apps/docs/pages/guides/functions/quickstart.mdx:91]:
```suggestion
Modify the `--data '{"name":"Functions"}'` line to `--data '{"name":"World"}'` and try invoking the command again.
```

File: apps/docs/pages/guides/functions/regional-invocation.mdx [hunk 27]
```diff
@@ -2,51 +2,54 @@ import Layout from '~/layouts/DefaultGuideLayout'
 
 export const meta = {
   id: 'function-regional-invocation',
-  title: 'Regional Invocation',
+  title: 'Regional Invocations',
   description: 'How to execute an Edge Functions in a particular region.',
+  subtitle: 'How to execute an Edge Functions in a particular region.',
```
Comment (charislam) [inline apps/docs/pages/guides/functions/regional-invocation.mdx:7]:
```suggestion
  subtitle: 'How to execute an Edge Function in a particular region.',
```

File: apps/docs/pages/guides/functions/regional-invocation.mdx [hunk 28]
```diff
@@ -2,51 +2,54 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
 }
 
-By default, Edge Functions are executed in the region closest to the user making the request. This helps to reduce network latency and provide faster responses to the user.
+Edge Functions are executed in the region closest to the user making the request. This helps to reduce network latency and provide faster responses to the user.
 
-However, if your function does lot of database or storage operations, invoking the function in the same region as your DB may provide better performance.
-
-Some situations where this might be helpful include:
+However, if your Function performs lot of database or storage operations, invoking the Function in the same region as your database may provide better performance. Some situations where this might be helpful include:
```
Comment (charislam) [inline apps/docs/pages/guides/functions/regional-invocation.mdx:12]:
```suggestion
However, if your Function performs lots of database or storage operations, invoking the Function in the same region as your database may provide better performance. Some situations where this might be helpful include:
```

File: apps/docs/pages/guides/functions/regional-invocation.mdx [hunk 29]
```diff
@@ -2,51 +2,54 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
 const { data, error } = await supabase.functions.invoke('hello-world', {
   body: { name: 'Functions' },
   headers: { 'x-region': 'eu-west-3' },
 })
 ```
 
-You can verify the region the function was executed by looking at the `x-sb-edge-region` HTTP header in the response. You can also find it as metadata in [Edge Function Logs](/docs/guides/functions/debugging)
+</CH.Code>
+
+You can verify the region the function was executed by looking at the `x-sb-edge-region` HTTP header in the response. You can also find it as metadata in [Edge Function Logs](/docs/guides/functions/debugging).
```
Comment (charislam) [inline apps/docs/pages/guides/functions/regional-invocation.mdx:50]:
```suggestion
You can verify the execution region by looking at the `x-sb-edge-region` HTTP header in the response. You can also find it as metadata in [Edge Function Logs](/docs/guides/functions/debugging).
```

**Review by charislam — APPROVED** [review]

File: apps/docs/pages/guides/functions/import-maps.mdx [hunk 30]
```diff
@@ -2,13 +2,13 @@ import Layout from '~/layouts/DefaultGuideLayout'
...
-  description: 'Managing packages using Import Maps.',
-  subtitle: 'Managing packages using Import Maps.',
+  title: 'Using Import Maps',
+  description: 'Managing packages and dependencies using Import Maps.',
+  subtitle: 'Managing packages and dependencies using Import Maps.',
   tocVideo: 'ILr3cneZuFk',
 }
 
-Import Maps are similar to a `package.json` file. They are a way to manage your dependencies in your Supabase Edge Functions.
+An Import Map us similar to a `package.json` file. They are a way to manage your dependencies.
```
Comment (laktek) [inline apps/docs/pages/guides/functions/import-maps.mdx]: An Import Map is

**Review by saltcod — APPROVED** [review]
